### PR TITLE
ath79: add support for NEC Aterm series based on QCA9558

### PR DESCRIPTION
**Pull request by musashino205**
This patch series adds support for NEC Aterm series devices based on Qualcomm Atheros QCA9558.

EDIT: U-Boot changes depend on https://github.com/openwrt/openwrt/pull/16703.

### CODE REVIEW

File: target/linux/ath79/dts/qca9558_nec_aterm.dtsi [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+		label-mac-device = &eth0;
+	};
+
+	i2c {
+		compatible = "i2c-gpio";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&pmx_gpio_i2c_pins>;
+		sda-gpios = <&gpio 20 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+		scl-gpios = <&gpio 19 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
```
Comment (DragonBluep) [inline target/linux/ath79/dts/qca9558_nec_aterm.dtsi:35]: ath79 hardware doesn't support open drain mode.

File: target/linux/ath79/dts/qca9558_nec_aterm.dtsi [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+			default-state = "on";
+		};
+	};
+
+	regulator {
+		compatible = "regulator-fixed";
+		regulator-name = "usb-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpio = <&gpio 22 GPIO_ACTIVE_HIGH>;
```
Comment (DragonBluep) [inline target/linux/ath79/dts/qca9558_nec_aterm.dtsi:180]: Please use `gpios`.

File: target/linux/ath79/dts/qca9558_nec_aterm.dtsi [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+	status = "okay";
+
+	nvmem-cells = <&macaddr_config_6>;
+	nvmem-cell-names = "mac-address";
+	pll-data = <0x03000000 0x00000101 0x00001616>;
+	qca955x-sgmii-fixup;
+
+	gmac-config {
+		device = <&gmac>;
+	};
```
Comment (DragonBluep) [inline target/linux/ath79/dts/qca9558_nec_aterm.dtsi:315]: This node can be removed I guess.

File: target/linux/ath79/dts/qca9558_nec_aterm.dtsi [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+&pinmux {
+	/* mux GPIO19/20 as GPIO instead of native I2C on QCA9558 */
+	pmx_gpio_i2c_pins: gpio-i2c-pins {
+		pinctrl-single,bits = <0x10 0x0 0xff000000>,
+				      <0x14 0x0 0xff>;
+	};
+};
+
+&uart0 {
+	status = "okay";
```
Comment (DragonBluep) [inline target/linux/ath79/dts/qca9558_nec_aterm.dtsi:351]: The default status is already set as “okay”.

File: target/linux/ath79/dts/qca955x.dtsi [hunk 5]
```diff
@@ -64,8 +64,8 @@
 				compatible ="qca,qca9550-usb-phy", "qca,ar7200-usb-phy";
 				reg = <0x18030000 4>, <0x18030004 4>;
 
-				reset-names = "usb-phy", "usb-suspend-override";
-				resets = <&rst 4>, <&rst 3>;
+				reset-names = "usb-phy-analog", "usb-phy", "usb-suspend-override";
+				resets = <&rst 11>, <&rst 4>, <&rst 3>;
```
Comment (DragonBluep) [inline target/linux/ath79/dts/qca955x.dtsi:68]: LGTM.

File: target/linux/ath79/dts/qca9558_nec_aterm.dtsi (hunk 1 above)
Comment (musashino205) [reply target/linux/ath79/dts/qca9558_nec_aterm.dtsi:35]: Oh, I see. I added `GPIO_OPEN_DRAIN` to resolve the warning messages printed by the kernel, but indeed, wrong binding should be removed.

File: target/linux/ath79/dts/qca9558_nec_aterm.dtsi (hunk 2 above)
Comment (musashino205) [reply target/linux/ath79/dts/qca9558_nec_aterm.dtsi:180]: OK, I'll update...

File: target/linux/ath79/dts/qca9558_nec_aterm.dtsi (hunk 3 above)
Comment (musashino205) [reply target/linux/ath79/dts/qca9558_nec_aterm.dtsi:315]: It is required to initialize SGMII on eth1 by `qca955x-sgmii-fixup;` property and `ag71xx_sgmii_init_qca955x()` function in ag71xx driver.

File: target/linux/ath79/dts/qca9558_nec_aterm.dtsi (hunk 1 above)
Comment (DragonBluep) [reply target/linux/ath79/dts/qca9558_nec_aterm.dtsi:35]: If there are some warnings I suggest keep it. At least it's harmless.

File: target/linux/ath79/dts/qca9558_nec_aterm.dtsi (hunk 4 above)
Comment (musashino205) [reply target/linux/ath79/dts/qca9558_nec_aterm.dtsi:351]: Oh, I didn't notice it. I'll remove...

File: target/linux/ath79/dts/qca9558_nec_aterm.dtsi (hunk 3 above)
Comment (DragonBluep) [reply target/linux/ath79/dts/qca9558_nec_aterm.dtsi:315]: But you already set it on `eth0`. Maybe you can try moving `gmac-config` from eth0 to eth1. I remember eth1 is probed first on ath79 target. If I'm not wrong these two `gmac-config` nodes configure the same registers.

**Comment by musashino205** [comment]
UPDATE:

- update QCA955x support patch in uboot-ath79 for v2024.10
  - replace `SUPPORTS_BIG_ENDIAN` to `SUPPORT_BIG_ENDIAN`
- use `gpios` in regulator node instead of `gpio`
- drop enabled `&uart0` node
- move `qca955x-sgmii-fixup` to eth0 node
- rebase from current main

U-Boot changes depend on https://github.com/openwrt/openwrt/pull/16703 now.

**Comment by musashino205** [comment]
UPDATE:

- remove a note about `SUPPORT_BIG_ENDIAN` from the U-Boot patch for adding QCA955x support
- rebase from the current main branch

**Pull request merged**